module/VuFind/src/VuFind/Auth/OpenIDConnect.php

```diff
@@ -0,0 +1,488 @@
+        }
+        return true;
+    }
+
+    /**
+     * Validate configuration parameters. This is a support method for getConfig(), so the configuration MUST be
+     * accessed using $this->oidcConfig; do not call $this->getConfig() from within this method!
+     *
+     * @throws AuthException
+     * @return void
+     */
+    protected function validateConfig(): void
+    {
+        $requiredParams = ['url', 'client_id', 'client_secret'];
+        foreach ($requiredParams as $param) {
+            if (empty($this->oidcConfig?->OpenIDConnect?->$param ?? null)) {
+                throw new AuthException(
+                    'One or more OpenID Connect parameters are missing. Check your OpenIDConnect.ini!'
+                );
+            }
+        }
+    }
+
+    /**
+     * Attempt to authenticate the current user. Throws exception if login fails.
+     *
+     * @param \Laminas\Http\PhpEnvironment\Request $request Request object containing account credentials.
+     *
+     * @throws AuthException
+     * @return User Object representing logged-in user.
+     */
+    public function authenticate($request): User
+    {
+        $code = $request->getQuery()->get('code');
+
+        if (empty($code)) {
+            throw new AuthException('authentication_error_admin');
+        }
+        $request_token = $this->getRequestToken($code);
+        $state = $request->getQuery()->get('state');
+        $stateIsValid = $state == $this->session->oidc_state;
+        unset($this->session->oidc_state);
+        if (!$stateIsValid) {
+            throw new AuthException('authentication_error_admin: bad state');
+        }
+
+        $claims = $this->decodeJWT($request_token->id_token);
+
+        if (!$this->validateIssuer($claims->iss)) {
+            throw new AuthException('authentication_error_admin: wrong issuer');
+        }
+        $claimsValid = $this->verifyJwtClaims($claims);
+        unset($this->session->oidc_nonce);
+        if (!$claimsValid) {
+            throw new AuthException('authentication_error: not valid claims');
+        }
+
+        $accessToken = $request_token->access_token;
+        $userInfo = $this->getUserInfo($accessToken);
+        return $this->setUserAttributes($userInfo);
+    }
+
+    /**
+     * Set user attributes from user info claim
+     *
+     * @param object $userInfo User info claim object
+     *
+     * @return User
+     * @throws AuthException
+     * @throws PasswordSecurityException
+     */
+    protected function setUserAttributes(object $userInfo): User
+    {
+        $availableAttributes = [
+            'firstname',
+            'lastname',
+            'email',
+            'cat_id',
+            'cat_username',
+            'cat_password',
+            'college',
+            'major',
+            'home_library',
+        ];
+        $user = $this->getUserTable()->getByUsername($userInfo->sub);
```

Comment:
I think we should have at least an option to prefix the sub with something that'd ensure it's unique and couldn't get mixed with e.g. Database authentication (e.g. when using ChoiceAuth).